### ISSUE TITLE
failure: PrefixAndTail must work even if tail subscriber arrives after substream completion

### ISSUE THREAD

**Issue by raboof** [akka/akka#23272]
https://jenkins.akka.io:8498/job/akka-nightly-2.12/254/testReport/junit/akka.stream.scaladsl/FlowPrefixAndTailSpec/PrefixAndTail_must_work_even_if_tail_subscriber_arrives_after_substream_completion/

```
[info] - must work even if tail subscriber arrives after substream completion *** FAILED *** (2 seconds, 36 milliseconds)
[info]   java.lang.IllegalStateException: Shutting down because of violation of the Reactive Streams specification.
[info]   at akka.stream.impl.fusing.GraphInterpreterShell.tryAbort(ActorGraphInterpreter.scala:600)
[info]   at akka.stream.impl.fusing.ActorGraphInterpreter.akka$stream$impl$fusing$ActorGraphInterpreter$$processEvent(ActorGraphInterpreter.scala:727)
[info]   at akka.stream.impl.fusing.ActorGraphInterpreter$$anonfun$receive$1.applyOrElse(ActorGraphInterpreter.scala:740)
```

**Comment by agolubev** [comment]
wow can reproduce it locally:
```
Cause: akka.stream.impl.ReactiveStreamsCompliance$SignalThrewException: It is illegal to throw exceptions from request(), rule 3.16
...
Cause: java.lang.NullPointerException:
```

**Comment by agolubev** [comment]
Ok, the issue I can explain on code snippet of test:
```
      val f = Source.fromPublisher(pub).prefixAndTail(1).runWith(Sink.head)
      val s = pub.expectSubscription()
      s.sendNext(0)
```
s here is PublisherProbeSubscription with BatchingActorInputBoundary inside
System does not wait while BatchingActorInputBoundary is initialized with subscriber.
So at line 3 `s.sendNext` it is race condition between onSubscribe and onNext in BatchingActorInputBoundary. In case onNext is faster we are seeing null pointer exception.

Per my understanding, we don't need to share subscription via pub.expectSubscription() - until it has been initialized.

**Comment by agolubev** [comment]
@ktoso, is there any existing class that is making all onNext events wait until finishing subscription?

**Comment by ktoso** [comment]
Great catch, thanks - no, I don't think we have a reusable wrapper for this behavior

**Comment by 2m** [comment]
Fixed in #23365
